File: src-frontend/services/brushDynamics.ts
/**
 * brushDynamics.ts
 * 
 * TypeScript bindings for K_OS Rust brush dynamics.
 * High-performance: interpolation, symmetry, and velocity grid.
 */

// Check if we're running in Tauri
const isTauri = (): boolean => {
    return typeof window !== 'undefined' && '__TAURI__' in window;
};

// Lazy import Tauri invoke
let tauriInvoke: ((cmd: string, args?: any) => Promise<any>) | null = null;

const getInvoke = async () => {
    if (tauriInvoke) return tauriInvoke;
    if (isTauri()) {
        try {
            const tauri = await import('@tauri-apps/api/core');
            tauriInvoke = tauri.invoke;
            return tauriInvoke;
        } catch (e) {
            console.warn('[brushDynamics] Failed to import Tauri API:', e);
        }
    }
    return null;
};

// ============================================================================
// TYPES
// ============================================================================

export interface InterpolatedPoint {
    uv: [number, number];
    pressure: number;
    t: number;
}

export interface SymmetryConfig {
    mirror_x: boolean;
    mirror_y: boolean;
    mirror_z: boolean;
    radial_count: number;
    center: [number, number];
}

export interface PaintOp {
    uv: [number, number];
    pressure: number;
}

// ============================================================================
// BRUSH INTERPOLATION - Replaces 100-raycast JS loop!
// ============================================================================

/**
 * Interpolate brush stroke between two UV positions.
 * Returns max 50 intermediate points (capped for performance).
 */
export async function interpolateBrushStroke(
    startUv: [number, number],
    endUv: [number, number],
    startPressure: number,
    endPressure: number,
    spacing: number,
    brushSizeUv: number
): Promise<InterpolatedPoint[]> {
    const invoke = await getInvoke();
    if (!invoke) {
        // Fallback: simple JS interpolation
        const dist = Math.hypot(endUv[0] - startUv[0], endUv[1] - startUv[1]);
        const steps = Math.min(Math.floor(dist / (brushSizeUv * spacing)), 20);
        const points: InterpolatedPoint[] = [];
        for (let i = 0; i <= steps; i++) {
            const t = i / Math.max(steps, 1);
            points.push({
                uv: [
                    startUv[0] + (endUv[0] - startUv[0]) * t,
                    startUv[1] + (endUv[1] - startUv[1]) * t
                ],
                pressure: startPressure + (endPressure - startPressure) * t,
                t
            });
        }
        return points;
    }

    try {
        return await invoke('interpolate_brush_stroke', {
            startUv,
            endUv,
            startPressure,
            endPressure,
            spacing,
            brushSizeUv,
        });
    } catch (e) {
        console.error('[brushDynamics] Interpolation failed:', e);
        return [{ uv: endUv, pressure: endPressure, t: 1 }];
    }
}

// ============================================================================
// SYMMETRY - Replaces triple JS loop!
// ============================================================================

/**
 * Calculate all symmetry UVs for a single paint point.
 * Replaces the triple JS mirror/radial loop.
 */
export async function calculateSymmetryUvs(
    uv: [number, number],
    config: SymmetryConfig
): Promise<[number, number][]> {
    const invoke = await getInvoke();
    if (!invoke) {
        // Fallback: simple JS symmetry
        const results: [number, number][] = [uv];
        if (config.mirror_x) results.push([1 - uv[0], uv[1]]);
        if (config.mirror_y) {
            const copy = [...results] as [number, number][];
            copy.forEach(p => results.push([p[0], 1 - p[1]]));
        }
        return results;
    }

    try {
        return await invoke('calculate_symmetry_uvs', { uv, config });
    } catch (e) {
        console.error('[brushDynamics] Symmetry failed:', e);
        return [uv];
    }
}

/**
 * Batch calculate symmetry for multiple points (parallel in Rust).
 */
export async function batchSymmetryUvs(
    uvs: [number, number][],
    config: SymmetryConfig
): Promise<[number, number][][]> {
    const invoke = await getInvoke();
    if (!invoke) {
        // Fallback
        return Promise.all(uvs.map(uv => calculateSymmetryUvs(uv, config)));
    }

    try {
        return await invoke('batch_symmetry_uvs', { uvs, config });
    } catch (e) {
        console.error('[brushDynamics] Batch symmetry failed:', e);
        return uvs.map(uv => [uv]);
    }
}

// ============================================================================
// VELOCITY GRID RGBA - No JS conversion needed!
// ============================================================================

/**
 * Get velocity grid as RGBA Float32Array ready for WebGL.
 * No JS conversion loop needed - Rust returns [R=vx, G=vy, B=0, A=1, ...].
 */
export async function getVelocityGridRgba(
    simId: number,
    resolution: number
): Promise<Float32Array | null> {
    const invoke = await getInvoke();
    if (!invoke) return null;

    try {
        const rgba: number[] = await invoke('get_velocity_grid_rgba', {
            simId,
            resolution,
        });
        return new Float32Array(rgba);
    } catch (e) {
        console.error('[brushDynamics] RGBA velocity failed:', e);
        return null;
    }
}

// ============================================================================
// FULL STROKE PROCESSING - Interpolation + Symmetry in ONE call!
// ============================================================================

/**
 * Process entire brush stroke: interpolate + symmetry in one Rust call.
 * Returns all paint operations ready for GPU batch.
 */
export async function processBrushStroke(
    startUv: [number, number],
    endUv: [number, number],
    startPressure: number,
    endPressure: number,
    spacing: number,
    brushSizeUv: number,
    symmetry?: SymmetryConfig
): Promise<PaintOp[]> {
    const invoke = await getInvoke();
    if (!invoke) {
        // Fallback: chain JS calls
        const points = await interpolateBrushStroke(
            startUv, endUv, startPressure, endPressure, spacing, brushSizeUv
        );
        if (!symmetry) {
            return points.map(p => ({ uv: p.uv, pressure: p.pressure }));
        }
        const allOps: PaintOp[] = [];
        for (const p of points) {
            const symUvs = await calculateSymmetryUvs(p.uv, symmetry);
            for (const uv of symUvs) {
                allOps.push({ uv, pressure: p.pressure });
            }
        }
        return allOps;
    }

    try {
        return await invoke('process_brush_stroke', {
            startUv,
            endUv,
            startPressure,
            endPressure,
            spacing,
            brushSizeUv,
            symmetry,
        });
    } catch (e) {
        console.error('[brushDynamics] Stroke processing failed:', e);
        return [{ uv: endUv, pressure: endPressure }];
    }
}

export default {
    interpolateBrushStroke,
    calculateSymmetryUvs,
    batchSymmetryUvs,
    getVelocityGridRgba,
    processBrushStroke,
};

File: crates/k-os-sculpt/src/brush_dynamics.rs
#![cfg(not(target_arch = "wasm32"))]
//! Brush Dynamics Module for KPainter
//!
//! High-performance brush calculations offloaded from JS:
//! - Batch raycast interpolation
//! - SIMD symmetry calculations
//! - Velocity grid in RGBA format (no JS conversion)

use glam::Vec2;
use rayon::prelude::*;
use serde::{Deserialize, Serialize};

// ============================================================================
// BRUSH INTERPOLATION - Replace JS loop with parallel Rust
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InterpolatedPoint {
    pub uv: [f32; 2],
    pub pressure: f32,
    pub t: f32, // Position along stroke (0-1)
}

/// Interpolate brush stroke between two UV positions
/// Returns all intermediate points for batch painting
#[tauri::command]
pub fn interpolate_brush_stroke(
    start_uv: [f32; 2],
    end_uv: [f32; 2],
    start_pressure: f32,
    end_pressure: f32,
    spacing: f32, // 0-1, fraction of brush size between dabs
    brush_size_uv: f32,
) -> Vec<InterpolatedPoint> {
    let start = Vec2::new(start_uv[0], start_uv[1]);
    let end = Vec2::new(end_uv[0], end_uv[1]);

    let dist = start.distance(end);
    let step_dist = (brush_size_uv * spacing).max(0.001);
    let steps = ((dist / step_dist) as usize).min(50); // Cap at 50 for performance

    if steps == 0 {
        return vec![InterpolatedPoint {
            uv: end_uv,
            pressure: end_pressure,
            t: 1.0,
        }];
    }

    (0..=steps)
        .map(|i| {
            let t = i as f32 / steps as f32;
            let uv = start.lerp(end, t);
            let pressure = start_pressure + (end_pressure - start_pressure) * t;
            InterpolatedPoint {
                uv: [uv.x, uv.y],
                pressure,
                t,
            }
        })
        .collect()
}

// ============================================================================
// SYMMETRY CALCULATIONS - SIMD-accelerated mirror/radial
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SymmetryConfig {
    pub mirror_x: bool,
    pub mirror_y: bool,
    pub mirror_z: bool,
    pub radial_count: u32,
    pub center: [f32; 2], // UV center for radial symmetry
}

/// Generate all symmetry UVs for a single paint point
/// Replaces the triple JS loop with a single Rust call
#[tauri::command]
pub fn calculate_symmetry_uvs(uv: [f32; 2], config: SymmetryConfig) -> Vec<[f32; 2]> {
    let mut results = Vec::with_capacity(16);
    let center = Vec2::new(config.center[0], config.center[1]);
    let point = Vec2::new(uv[0], uv[1]);

    // Start with original point
    let mut base_points = vec![point];

    // Mirror X
    if config.mirror_x {
        let mirrored: Vec<_> = base_points
            .iter()
            .map(|p| Vec2::new(1.0 - p.x, p.y))
            .collect();
        base_points.extend(mirrored);
    }

    // Mirror Y
    if config.mirror_y {
        let mirrored: Vec<_> = base_points
            .iter()
            .map(|p| Vec2::new(p.x, 1.0 - p.y))
            .collect();
        base_points.extend(mirrored);
    }

    // Radial symmetry
    if config.radial_count > 1 {
        let angle_step = std::f32::consts::TAU / config.radial_count as f32;
        for base in &base_points.clone() {
            let offset = *base - center;
            for i in 1..config.radial_count {
                let angle = angle_step * i as f32;
                let cos_a = angle.cos();
                let sin_a = angle.sin();
                let rotated = Vec2::new(
                    offset.x * cos_a - offset.y * sin_a + center.x,
                    offset.x * sin_a + offset.y * cos_a + center.y,
                );
                base_points.push(rotated);
            }
        }
    }

    // Convert to output format, filtering out-of-bounds
    for p in base_points {
        if p.x >= 0.0 && p.x <= 1.0 && p.y >= 0.0 && p.y <= 1.0 {
            results.push([p.x, p.y]);
        }
    }

    results
}

/// Batch calculate symmetry for multiple points (for stroke interpolation)
#[tauri::command]
pub fn batch_symmetry_uvs(uvs: Vec<[f32; 2]>, config: SymmetryConfig) -> Vec<Vec<[f32; 2]>> {
    uvs.into_par_iter()
        .map(|uv| calculate_symmetry_uvs(uv, config.clone()))
        .collect()
}

// ============================================================================
// VELOCITY GRID - Return RGBA directly (no JS conversion!)
// ============================================================================

use lazy_static::lazy_static;
use std::collections::HashMap;
use std::sync::Mutex;

lazy_static! {
    // Pre-allocated RGBA buffers per resolution to avoid allocations
    static ref RGBA_BUFFERS: Mutex<HashMap<usize, Vec<f32>>> = Mutex::new(HashMap::new());
}

/// Get velocity grid as RGBA (no JS conversion needed!)
/// Returns [R=vx, G=vy, B=0, A=1, ...] ready for WebGL upload
#[cfg(feature = "physics")]
#[tauri::command]
pub fn get_velocity_grid_rgba(sim_id: u64, resolution: usize) -> Result<Vec<f32>, String> {
    use k_os_sim::FLUID_SIMS;

    let sims = FLUID_SIMS.lock().map_err(|e| e.to_string())?;
    let sim = sims.get(&sim_id).ok_or("Simulation not found")?;

    // Get velocity grid [vx, vy, vx, vy, ...]
    let velocity = sim.get_velocity_grid(resolution);

    // Convert to RGBA in parallel
    let pixel_count = resolution * resolution;
    let rgba: Vec<f32> = (0..pixel_count)
        .into_par_iter()
        .flat_map(|i| {
            let vx = velocity.get(i * 2).copied().unwrap_or(0.0);
            let vy = velocity.get(i * 2 + 1).copied().unwrap_or(0.0);
            vec![vx, vy, 0.0, 1.0]
        })
        .collect();

    Ok(rgba)
}

// ============================================================================
// BATCH PAINT OPERATIONS - Full stroke processing in Rust
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PaintOp {
    pub uv: [f32; 2],
    pub pressure: f32,
}

/// Process entire brush stroke: interpolate + symmetry in one call
/// Returns all paint operations ready for GPU batch
#[tauri::command]
pub fn process_brush_stroke(
    start_uv: [f32; 2],
    end_uv: [f32; 2],
    start_pressure: f32,
    end_pressure: f32,
    spacing: f32,
    brush_size_uv: f32,
    symmetry: Option<SymmetryConfig>,
) -> Vec<PaintOp> {
    // 1. Interpolate
    let points = interpolate_brush_stroke(
        start_uv,
        end_uv,
        start_pressure,
        end_pressure,
        spacing,
        brush_size_uv,
    );

    // 2. Apply symmetry if enabled
    let mut all_ops = Vec::with_capacity(points.len() * 8);

    if let Some(sym) = symmetry {
        for p in points {
            let sym_uvs = calculate_symmetry_uvs(p.uv, sym.clone());
            for uv in sym_uvs {
                all_ops.push(PaintOp {
                    uv,
                    pressure: p.pressure,
                });
            }
        }
    } else {
        for p in points {
            all_ops.push(PaintOp {
                uv: p.uv,
                pressure: p.pressure,
            });
        }
    }

    all_ops
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_interpolation() {
        let points = interpolate_brush_stroke([0.0, 0.0], [1.0, 1.0], 1.0, 0.5, 0.1, 0.1);
        assert!(points.len() > 1);
        assert!(points.len() <= 51);
    }

    #[test]
    fn test_symmetry() {
        let config = SymmetryConfig {
            mirror_x: true,
            mirror_y: false,
            mirror_z: false,
            radial_count: 1,
            center: [0.5, 0.5],
        };
        let uvs = calculate_symmetry_uvs([0.25, 0.5], config);
        assert_eq!(uvs.len(), 2); // Original + X mirror
    }
}
